crates/ruma-events/src/direct.rs

```diff
@@ -80,20 +139,43 @@ mod tests {
 
     #[test]
     fn deserialization() {
-        let alice = owned_user_id!("@alice:ruma.io");
+        let alice = DirectUserIdentifier::from_borrowed("@alice:ruma.io");
         let rooms = vec![owned_room_id!("!1:ruma.io"), owned_room_id!("!2:ruma.io")];
 
         let json_data = json!({
             "content": {
-                alice.to_string(): rooms,
+                alice: rooms,
+                "alice@ruma.io": vec![owned_room_id!("!3:ruma.io")],
             },
             "type": "m.direct"
         });
 
         let event: DirectEvent = from_json_value(json_data).unwrap();
-        let direct_rooms = event.content.get(&alice).unwrap();
+        let direct_rooms = event.content.get(alice).unwrap();
 
         assert!(direct_rooms.contains(&rooms[0]));
         assert!(direct_rooms.contains(&rooms[1]));
     }
+
+    #[test]
+    fn user_id_conversion() {
+        let alice_direct_uid = DirectUserIdentifier::from_borrowed("@alice:ruma.io");
+        let alice_owned_user_id: OwnedUserId = alice_direct_uid
+            .to_owned()
+            .try_into()
+            .expect("@alice:ruma.io should be convertible into a Matrix user ID");
+        assert_eq!(alice_direct_uid.as_str(), alice_owned_user_id.as_str());
```

Comment:
For ease of use, it would probably make sense to implement `PartialEq` between the `UserId`and `DirectUserIdentifier` types? i.e.:

```rust
impl PartialEq<&UserId> for &DirectUserIdentifier {}
impl PartialEq<OwnedUserId> for &DirectUserIdentifier {}
impl PartialEq<&UserId> for OwnedDirectUserIdentifier {}
impl PartialEq<OwnedUserId> for OwnedDirectUserIdentifier {}
```

And vice-versa.